hubble: Show applied L4 network policies in hubble drop event

41ks opened the pull request:
Adds option to show blocking or applied network policies in `hubble` k8s drop events. This allows for faster diagnostic of this type of event. Also, adds rate limiting for the drop event emitter to limit the resource usage of this feature.

The feature has been placed behind a flag for now. But, in our experience it can be very useful. So it could be turned on by default or the flag could be removed altogether.

This is also intended to help move this feature to beta #33975.

## How it works

When a drop event occurs, the correlation results are parsed to detect if the drop was due to a specific deny policy. If the correlation results are empty, all applied network policies are shown. The emitter will query the endpoint manager to retrieve the applied L4 rules. Those rules are then parsed and shown as either namespaced or clusterwide network policies.

## Result

The following is an example of a dropevent with the feature turned off:
```
Outgoing packet dropped (policy_denied) to 8.8.8.8 ICMPv4.
Incoming packet dropped (policy_denied) from remote-node (1.2.3.4) TCP/8080.
```

With the feature turned off we get something like this:
```
Outgoing packet dropped (policy_deny) to 8.8.8.8 ICMPv4. Denied by policies: deny-icmp.
Outgoing packet dropped (policy_denied) to 8.8.8.8 ICMPv4. Applied policies: namespaced-cnp. Applied clusterwide policies: cluster-cnp-1, cluster-cnp-2.
Incoming packet dropped (policy_denied) from remote-node (1.2.3.4) TCP/8080. Applied policies: namespaced-cnp. Applied clusterwide policies: cluster-cnp-1, cluster-cnp-2.
```

## Performance Analysis

To assess the CPU impact without rate limiting, we deployed this change on one of our machines and subjected it to deliberate packet drops with varying rates.

The values in the table below correspond to average CPU usage in `mcores`. Idle CPU core usage = around `58 mcores`

| Configuration | 100 pkt/s | 500 pkt/s | 1000 pkt/s | Event message |
|--------|--------|--------|--------|--------|
| No events | 102.62 | 111.42 | 152.42 | N/A |
| Dropevents | 106.63 | 178.82 | 275.13 | Outgoing packet dropped (policy_denied) to 8.8.8.8 ICMPv4. |
| Dropevents + extended (Policy Deny) | 108.9 | 166.92 | 253.15 | Outgoing packet dropped (policy_deny) to 8.8.4.4 ICMPv4. Denied by policies: alex-custom-cnp. |
| Dropevents + extended (Policy Denied) | 136.66 | 253.74 | 430.9 | Outgoing packet dropped (policy_denied) to 8.8.8.8 ICMPv4. Applied policies: alex-custom-cnp. Applied clusterwide policies: ... |

This shows a slight impact from this feature. However, the packet drop rates used here exceed those typically observed on production machines under normal conditions.

With rate limiting turned on, performance is stable across packet drop rates.

## Release note
```release-note
Hubble v1.Events drop message now supports displaying applied L4 network policies
```

bimmlerd reviewed (changes requested):
Haven't reviewed in detail yet - one question first: why does this not rely on the existing correlation function (in `policy/correlation`)?

joamaki reviewed (approved):
`pkg/k8s` changes LGTM

bimmlerd commented on pkg/hubble/dropeventemitter/cell.go:
```diff
@@ -89,6 +94,8 @@ type params struct {
 
 	Config config
+
+	EndpointManager endpointmanager.EndpointManager
```
I don't think you need the full EPMgr interface, EndpointsLookup should be enough

41ks replied:
Thanks for the tip I'll fix it

bimmlerd commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -147,3 +177,93 @@ func l4protocolToString(l4 *flowpb.Layer4) string {
+		return nil, 0
+	}
+
+	model := ep.GetModel().Status
```
`GetModel` seems like overkill for what you're wanting to do - it seems pretty expensive to be doing this per flow... Are you running this in prod already/do you have an understanding on how much this costs?

41ks replied:
There is a method on the endpoint object `GetPolicyModel()` which returns only the part of the model which is used here. However, this method needs to be called with e.mutex Rlock()ed. I could move this method to `getPolicyModel` and expose an Rlock()ed method `GetPolicyModel()`. Or I can create a new method which sends the bare minimum for my logic to work.

bimmlerd reviewed (changes requested):
How does this interact with the default deny setting? It seems to me with defautlDeny on, you ought to somehow emit that a flow was dropped because no policy applied, whereas with it off there must be a specific policy which caused the drop

devodev commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -69,6 +72,37 @@ func GetPolicyLabels(ns, name string, uid types.UID, derivedFrom string) labels.
 	return append(labelsArr, srcLabel)
 }
 
+// GetPolicyFromLabels derives and sets fields in the flow policy from the label set array
```
```suggestion
// GetPolicyFromLabels derives and sets fields in the flow policy from the label set array.
```

devodev commented on pkg/policy/l4.go:
```diff
@@ -1813,6 +1813,43 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 	}
 }
 
+func (l4 *L4Policy) GetRuleOriginModel() *models.L4Policy {
```
Can we add a doc comment? And while we're here, can we do `GetModel()` as well?

devodev commented on pkg/endpoint/api.go:
```diff
@@ -647,3 +647,19 @@ func (e *Endpoint) ApplyUserLabelChanges(lbls labels.Labels) (add, del labels.La
 }
+
+// GetRealizedL4PolicyModel returns the realized L4 policy of the endpoint.
+func (e *Endpoint) GetRealizedL4PolicyRuleOriginModel() (policy *models.L4Policy, policyRevision uint64, err error) {
```
```suggestion
// GetRealizedL4PolicyRuleOriginModel returns the realized L4 policy of the endpoint.
func (e *Endpoint) GetRealizedL4PolicyRuleOriginModel() (policy *models.L4Policy, policyRevision uint64, err error) {
```
PR has been turned to draft but just want to make sure we don't forget to update names.

devodev commented on pkg/hubble/dropeventemitter/cell.go:
```diff
@@ -41,6 +42,8 @@ type config struct {
 	// K8sDropEventsReasons controls which drop reasons to emit events for.
 	K8sDropEventsReasons []string `mapstructure:"hubble-drop-events-reasons"`
+	// EnableK8sDropEventsPolicies controls if L4 network policies are included in event message
+	EnableK8sDropEventsExtended bool `mapstructure:"hubble-drop-events-extended"`
```
```suggestion
	// EnableK8sDropEventsExtended controls if L4 network policies are included in event message.
	EnableK8sDropEventsExtended bool `mapstructure:"hubble-drop-events-extended"`
```

devodev replied:
nit: Still missing the dot at the end.

devodev commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -147,3 +183,85 @@ func l4protocolToString(l4 *flowpb.Layer4) string {
+	return e.endpointsLookup.LookupCiliumID(endpointID)
+}
+
+func getPolicyRulesFromEndpoint(direction flowpb.TrafficDirection, ep endpointInterface) ([]*models.PolicyRule, uint64, error) {
+	if ep == nil {
+		return nil, 0, nil
+	}
+	realizedPolicy, policyRevision, err := ep.GetRealizedL4PolicyRuleOriginModel()
+	if err != nil || realizedPolicy == nil {
+		return nil, 0, err
+	}
+	if direction == flowpb.TrafficDirection_INGRESS {
+		return realizedPolicy.Ingress, policyRevision, nil
+	}
+	return realizedPolicy.Egress, policyRevision, nil
+}
```
Could we add test cases for error handling cases, such as when ep is nil ?

devodev reviewed (commented):
Mostly highlighting some nits to not forget before merging.

gandro commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -147,3 +183,85 @@ func l4protocolToString(l4 *flowpb.Layer4) string {
+	var parts []string
+	prefix := "Denied by"
+	networkPolicies, clusterwideNetworkPolicies := parsePolicyCorrelation(flow.TrafficDirection, flow.IngressDeniedBy, flow.EgressDeniedBy)
+	if networkPolicies.Len() == 0 && clusterwideNetworkPolicies.Len() == 0 {
```
optional nit: I believe this logic could be further optimized by looking at the drop reason:

1. If the `DropReasonDesc` is `DropReason_POLICY_DENIED`, then we hit a default-deny. Which means we collect all applied policies from the endpoint
2. If the `DropReasonDesc` is `DropReason_POLICY_DENY` (great wording, I know), then we hit a deny policy. This means we should be able to extract matched policies from policy correlation.

41ks replied:
I have made changes to implement your suggestion. I still prefer to check if the sets are empty instead of relying only on the drop reason description since policy correlation could be turned off by the user.

gandro replied:
Yeah, good call with the policy feature being disabled! Looks good to me now

gandro commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -147,3 +183,85 @@ func l4protocolToString(l4 *flowpb.Layer4) string {
+	return realizedPolicy.Egress, policyRevision, nil
+}
+
+func parsePolicyRules(l4Rules []*models.PolicyRule, policyRevision uint64) (networkPolicies set.Set[string], clusterwideNetworkPolicies set.Set[string]) {
```
bug: This seems to assume that only CiliumNetworkPolicy (CNP) and ClusterwideCiliumNetworkPolicy (CCNP) exist. However, there is also Kubernetes NetworkPolicy (KNP) and (soon, PR not yet merged) Kubernetes ClusterNetworkPolicy (KCNP).

If you have a KNP named "default/test" and a CNP named "default/test", it will only report one of the two, even though both exist. You should consider reporting the policy kind (found in `flow.Policy.Kind` too).

41ks replied:
I have added the network policy kind in the name of the policies to be included in the event message. This should solve this bug and it also provides users with more information about the source of a drop.

gandro commented on pkg/policy/l4.go:
```diff
@@ -1813,6 +1814,44 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 }
 
+// GetRuleOriginModel returns the API model of the L4 policy with the rule origins only.
+func (l4 *L4Policy) GetRuleOriginModel() *models.L4Policy {
```
optional suggestion: I don't have a concrete proposal here, but it seems like an awfully expensive function doing lots of allocaiton for something which essentially only exists to extract the network policy name from its "DerivedFrom" labels.

I think it would be potentially more efficient to move a bit of the code in here and essentially just return a list of "policykind/policynamespace/policyname" structs for both ingress and egress here, instead of returning a fully populated model.

41ks replied:
From the CPU usage profiling I ran, the most resource-intensive tasks appear to be related to label handling. I’m not sure how to optimize this further, but I’m open to suggestions. As for memory usage, the feature hasn’t shown any significant impact so far, but I can certainly work on improving memory allocations if needed.

gandro replied:
Thanks for looking into this! Yeah I don't think we need to address this unless it becomes a problem. The current approach certainly looks clean

gandro reviewed (commented):
Thanks a lot and apologies for the delay. I spotted one bug, the rest are optional suggestions considering this feature is opt-in.

Also note that you likely need to rebase this branch on current main to make sure you've pulled in the latest policy refactor.

gandro reviewed (approved):
Nice, thank you!

bimmlerd reviewed (approved):
Thanks!

kaworu reviewed (commented):
Thank you @41ks for the PR, and sorry for the late review.

Functionally the patch LGTM. I have two performance related questions:
1. Related to [promoting the Hubble drop emitter to beta](https://github.com/cilium/cilium/issues/33975), I see that the `dropEventEmitter` configures its k8s client with `CorrelatorOptions` so that we have a limited amount of event sent per second. That's good, but I think this patch introduces enough overhead to ponder whether we should handle rate-limiting earlier, i.e. before doing any kind of policy correlation. Otherwise it seems like we're doing a lot of extra work that end up dropped by the k8s client rate-limiting. In the Performance Analysis section of this PR description, it is striking that overhead increases as more packet are dropped, because if I understand correctly we get the same number of k8s events in all cases (since the default interval is 2 minutes).
2. I think there should be a "safety valve" on the count of policy handled. In your examples there are a handful of policies, but some environments could have orders of magnitude more. Arguably, there is a cap of how helpful the full policy list can be (in the default deny case), so the extreme case is both resource intensive and not helpful.

What do you think?

41ks committed:
hubble: Add rate limiting to the drop event emitter

This adds rate limiting to drop event messages. This allows the user to limit the performance impact of this feature.

Signed-off-by: Alex Melhem <alex.melhem@datadoghq.com>

41ks committed:
hubble: Show blocking or applied network policies in hubble drop event

Adds option to show blocking or applied network policies in hubble k8s drop event. This allows for faster diagnostic of this type of event.

Signed-off-by: Alex Melhem <alex.melhem@datadoghq.com>

kaworu reviewed (approved):
Thanks for the quick update, patch LGTM.

The rate-limit capability is welcome, it would be great to improve the UX (i.e. flags) by consolidating the event interval and rate-limiting into one (as-is I find it a bit confusing for users without knowledge of the internals of creating events vs emitting events). I don't want to block this PR so let's merge as-is and take note for future work.

The "too many policies" improvement can be done later on as well.

I've added both comments to https://github.com/cilium/cilium/issues/33975.
